tests/vxlan/test_vnet_route_leak.py

```diff
@@ -0,0 +1,188 @@
+import logging
+import pytest
+import re
+
+from tests.common.helpers.assertions import pytest_assert
+from tests.common.utilities import wait_until
+from vnet_constants import *
+from vnet_utils import cleanup_vnet_routes, cleanup_dut_vnets, cleanup_vxlan_tunnels, \
+                       apply_dut_config_files, generate_dut_config_files
+
+logger = logging.getLogger(__name__)
+
+pytestmark = [
+    pytest.mark.topology("t0")
+]
+
+BGP_WAIT_TIMEOUT = 120
+
+TESTING_STATUS = "Testing"
+CLEANUP_STATUS = "Cleanup"
+
+SHOW_VNET_ROUTES_CMD = "show vnet routes all"
+SHOW_BGP_SUMMARY_CMD = "show ip bgp summary"
+SHOW_BGP_ADV_ROUTES_CMD_TEMPLATE = "show ip bgp neighbor {} advertised-routes"
+RESTART_BGP_CMD = "sudo systemctl restart bgp"
+CONFIG_SAVE_CMD = "sudo config save -y"
+CONFIG_RELOAD_CMD = "sudo config reload -y"
+BACKUP_CONFIG_DB_CMD = "sudo cp /etc/sonic/config_db.json /etc/sonic/config_db.json.route_leak_orig"
+RESTORE_CONFIG_DB_CMD = "sudo cp /etc/sonic/config_db.json.route_leak_orig /etc/sonic/config_db.json"
+DELETE_BACKUP_CONFIG_DB_CMD = "sudo rm /etc/sonic/config_db.json.route_leak_orig"
+
+BGP_ERROR_TEMPLATE = "BGP sessions not established after {} seconds"
+LEAKED_ROUTES_TEMPLATE = "Leaked routes: {}"
+
+
+@pytest.fixture(scope="module")
+def configure_dut(minigraph_facts, duthost, vnet_config, vnet_test_params):
+    """
+    Setup/teardown fixture for VNET route leak test
+
+    During the setup portion, generates VNET VxLAN configurations and applies them to the DUT
+    During the teardown portion, removes all previously pushed VNET VxLAN information from the DUT
+
+    Args:
+        minigraph_facts: Minigraph information
+        duthost: DUT host object
+        vnet_config: Dictionary containing VNET configuration information
+        vnet_test_params: Dictionary containing VNET test parameters
+    """
+
+    logger.info("Backing up config_db.json")
+    duthost.shell(BACKUP_CONFIG_DB_CMD)
+
+    duthost.shell("sonic-clear fdb all")
+    generate_dut_config_files(duthost, minigraph_facts, vnet_test_params, vnet_config)
+    apply_dut_config_files(duthost, vnet_test_params)
+
+    # In this case yield is used only to separate this fixture into setup and teardown portions
+    yield
+
+    if vnet_test_params[CLEANUP_KEY]:
+        logger.info("Restoring config_db.json")
+        duthost.shell(RESTORE_CONFIG_DB_CMD)
+        duthost.shell(DELETE_BACKUP_CONFIG_DB_CMD)
+
+        cleanup_vnet_routes(duthost, vnet_test_params)
+        cleanup_dut_vnets(duthost, minigraph_facts, vnet_config)
+        cleanup_vxlan_tunnels(duthost, vnet_test_params)
+
+        duthost.shell(RESTART_BGP_CMD)
```

Comment:
Rather than doing all of this, I would just `config reload` to load the original config DB that you restored above. It should cover these 4 steps and it has a stronger guarantee of leaving the successive test cases in a clean state.

Reply:
That's a good point, but do we care about how long it takes to run? `config reload` takes a while to finish but these steps finish relatively quickly.

Reply:
I'll defer to @yxieca here, but I think the current priority is stability over speed.

Reply:
Fixed

Reply:
I agree with @daall. But I am also open to try this limited recover since @theasianpianist knows this test better.

However, this recovery method is not complete. You shouldn't restart bgp and go, you should restart bgp, check the system restored, then go. If the system failed to restore, then you can invoke the config_reload fixture which knows waiting for the system to restore.

Reply:
@yxieca , @daall , this is part of a test to check if routes are leaked by bgp restart. May be we could wait for bgp sessions to be established before checking for leaked routes?

Reply:
@yxieca Would waiting for all the BGP sessions to be connected be sufficient? Or is there some other metric to indicate the system is ready?

Reply:
@prsunny this is clean up code. when reaching here. test is done.
@theasianpianist If BGP is only component impacted, then you can wait until BGP is up and default route is restored, then move on, but if timed out waiting for these, then you should issue config_reload to recover.

Reply:
oh, this part is already changed in the new commit.  As I see, there is no bgp restart anymore during cleanup.

Reply:
Thanks @theasianpianist, if the change is not a big one, go for it. Otherwise, you can always come back with another PR to address the issue since config_reload will be a sure recover. I leave this choice to you.

Reply:
Cool, resolving this thread then. Will update to use BGP restart and fallback to `config reload`.